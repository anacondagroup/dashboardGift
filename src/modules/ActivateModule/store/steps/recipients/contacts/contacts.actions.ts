import { createAction } from 'redux-act';
import { TErrors } from '@alycecom/services';

import { IContact, IContactsFilters, IContactsMeta } from './contacts.types';

const PREFIX = 'ACTIVATE_MODULE/STEPS/RECIPIENTS/CONTACTS';

export const loadContactsRequest = createAction<{ campaignId: number; reset?: boolean }>(`${PREFIX}/LOAD_REQUEST`);
export const loadContactsSuccess = createAction<{ data: IContact[]; meta: IContactsMeta }>(`${PREFIX}/LOAD_SUCCESS`);
export const loadContactsFail = createAction<TErrors>(`${PREFIX}/LOAD_FINISH`);

export const setContactsFilters = createAction<Partial<IContactsFilters>>(`${PREFIX}/SET_FILTERS`);
export const resetStatusMetaData = createAction(`${PREFIX}/RESET_REQUEST_STATUS_META_DATA`);
