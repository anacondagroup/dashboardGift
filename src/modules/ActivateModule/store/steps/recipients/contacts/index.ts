export { contacts } from './contacts.reducer';
export type { IContactsState } from './contacts.reducer';
export * from './contacts.actions';
export * from './contacts.selectors';
export * from './contacts.epics';
