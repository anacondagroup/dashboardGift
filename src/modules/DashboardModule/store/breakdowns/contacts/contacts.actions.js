import {
  CONTACTS_BREAKDOWN_LOAD_REQUEST,
  CONTACTS_BREAKDOWN_LOAD_SUCCESS,
  CONTACTS_BREAKDOWN_LOAD_FAILS,
  CONTACTS_DOWNLOAD_REPORT_REQUEST,
  CONTACTS_DOWNLOAD_REPORT_SUCCESS,
  CONTACTS_DOWNLOAD_REPORT_FAIL,
} from './contacts.types';

export const contactsLoadRequest = options => ({
  type: CONTACTS_BREAKDOWN_LOAD_REQUEST,
  payload: options,
});

export const contactsLoadSuccess = breakdown => ({
  type: CONTACTS_BREAKDOWN_LOAD_SUCCESS,
  payload: breakdown,
});

export const contactsLoadFail = error => ({
  type: CONTACTS_BREAKDOWN_LOAD_FAILS,
  payload: error,
});

export const contactsDownloadReportRequest = options => ({
  type: CONTACTS_DOWNLOAD_REPORT_REQUEST,
  payload: options,
});

export const contactsDownloadReportSuccess = () => ({
  type: CONTACTS_DOWNLOAD_REPORT_SUCCESS,
});

export const contactsDownloadReportFail = () => ({
  type: CONTACTS_DOWNLOAD_REPORT_FAIL,
});
