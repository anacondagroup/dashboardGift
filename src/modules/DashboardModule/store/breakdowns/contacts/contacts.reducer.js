import * as R from 'ramda';
import { renameKeys } from '@alycecom/utils';

import { initialBreakdownState } from '../../../helpers/breakdownSelectors.helpers';
import { setLoading, setLoaded } from '../../../../../helpers/lens.helpers';

import {
  CONTACTS_BREAKDOWN_LOAD_REQUEST,
  CONTACTS_BREAKDOWN_LOAD_SUCCESS,
  CONTACTS_BREAKDOWN_LOAD_FAILS,
  CONTACTS_DOWNLOAD_REPORT_REQUEST,
  CONTACTS_DOWNLOAD_REPORT_SUCCESS,
  CONTACTS_DOWNLOAD_REPORT_FAIL,
} from './contacts.types';

export const initialState = {
  ...initialBreakdownState,
};

export const renameBreakdownKeys = renameKeys({
  gifts_sent: 'giftsSent',
  gifts_viewed: 'giftsViewed',
  gifts_accepted: 'giftsAccepted',
  meetings_booked: 'meetingsBooked',
  last_contact: 'lastContact',
  full_name: 'fullName',
});

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case CONTACTS_BREAKDOWN_LOAD_REQUEST:
      return {
        ...state,
        ...setLoading(true, state),
      };
    case CONTACTS_BREAKDOWN_LOAD_SUCCESS:
      return {
        ...state,
        ...R.compose(setLoading(false), setLoaded(true))(state),
        breakdown: action.payload.map(recipient => renameBreakdownKeys(recipient)),
      };
    case CONTACTS_BREAKDOWN_LOAD_FAILS:
      return {
        ...state,
        ...setLoading(false, state),
        error: action.payload,
      };
    case CONTACTS_DOWNLOAD_REPORT_REQUEST:
      return {
        ...state,
        isReportLoading: true,
      };
    case CONTACTS_DOWNLOAD_REPORT_SUCCESS:
    case CONTACTS_DOWNLOAD_REPORT_FAIL:
      return {
        ...state,
        isReportLoading: false,
      };
    default:
      return state;
  }
};
