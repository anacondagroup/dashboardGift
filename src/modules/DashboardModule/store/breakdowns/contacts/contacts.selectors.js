import * as R from 'ramda';
import { updateSearch } from '@alycecom/modules';
import { pickNotEmpty, renameKeys } from '@alycecom/utils';

import { getBreakdownsState } from '../breakdowns.selectors';
import { createSelectors } from '../../../helpers/breakdownSelectors.helpers';

export const contactsApiParams = R.pick(['dateRangeFrom', 'dateRangeTo', 'sort', 'sortDirection', 'search']);

export const [
  getContactsState,
  getContactsBreakdown,
  getContactsIsLoading,
  getContactsReportIsLoading,
  getContactsIsLoaded,
] = createSelectors(getBreakdownsState)('contacts');

export const getApiCallQuery = R.compose(
  R.curryN(2, updateSearch)(''),
  renameKeys({
    sort: 'sort_column',
    sortDirection: 'sort_direction',
    search: 'search',
    dateRangeFrom: 'date_range_from',
    dateRangeTo: 'date_range_to',
  }),
  pickNotEmpty,
  contactsApiParams,
);

export const buildApiUrl = options => `/enterprise/dashboard/breakdown/recipients?${getApiCallQuery(options)}`;
export const buildReportUrl = options => `/enterprise/report/recipient_breakdown?${getApiCallQuery(options)}`;

export const getReportFileName = apiParams => () => {
  let date = '';

  if (apiParams.dateRangeFrom) {
    date = `-${apiParams.dateRangeFrom}`;
  }

  if (apiParams.dateRangeTo) {
    date += `-${apiParams.dateRangeTo}`;
  }

  return R.replace(/\./g, '_', `Contact breakdown${date || ''}`).concat('.xlsx');
};
