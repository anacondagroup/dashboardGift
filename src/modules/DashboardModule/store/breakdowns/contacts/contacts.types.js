export const CONTACTS_BREAKDOWN_LOAD_REQUEST = 'CONTACTS_BREAKDOWN_LOAD_REQUEST';
export const CONTACTS_BREAKDOWN_LOAD_SUCCESS = 'CONTACTS_BREAKDOWN_LOAD_SUCCESS';
export const CONTACTS_BREAKDOWN_LOAD_FAILS = 'CONTACTS_BREAKDOWN_LOAD_FAILS';

export const CONTACTS_DOWNLOAD_REPORT_REQUEST = 'CONTACTS_DOWNLOAD_REPORT_REQUEST';
export const CONTACTS_DOWNLOAD_REPORT_SUCCESS = 'CONTACTS_DOWNLOAD_REPORT_SUCCESS';
export const CONTACTS_DOWNLOAD_REPORT_FAIL = 'CONTACTS_DOWNLOAD_REPORT_FAIL';
