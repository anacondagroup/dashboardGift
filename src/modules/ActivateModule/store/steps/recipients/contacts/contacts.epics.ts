import { Epic } from 'redux-observable';
import { catchError, concatMap, distinctUntilChanged, filter, map, mergeMap, withLatestFrom } from 'rxjs/operators';
import { handlers } from '@alycecom/services';
import { ofType } from '@alycecom/utils';
import qs from 'query-string';

import { getActivateModuleParams } from '../../../activate.selectors';
import { pollContactsUploadingFileFinish } from '../uploadRequest';
import { ActivateModes } from '../../../../routePaths';

import {
  loadContactsFail,
  loadContactsRequest,
  loadContactsSuccess,
  resetStatusMetaData,
  setContactsFilters,
} from './contacts.actions';
import { IContactsResponse } from './contacts.types';
import { getContactsFilters, getContactsPagination, getMetaData } from './contacts.selectors';

const contactsQueryParamsSideEffectEpic: Epic = (action$, state$) =>
  action$.pipe(
    ofType(setContactsFilters),
    withLatestFrom(state$),
    map(([, state]) => {
      const { campaignId } = getActivateModuleParams(state);
      return loadContactsRequest({ campaignId });
    }),
  );

const loadContactsEpic: Epic = (action$, state$, { apiService }) =>
  action$.pipe(
    ofType(loadContactsRequest),
    withLatestFrom(state$),
    map(([action, state]) => {
      const {
        payload: { campaignId },
      } = action;
      const { search, sortDirection, sortField } = getContactsFilters(state);
      const { limit, offset } = getContactsPagination(state);
      const meta = getMetaData(state);

      const params = {
        limit,
        search,
        offset,
        sort: `${sortField}:${sortDirection.toLowerCase()}`,
      };
      const queryString = qs.stringify(params);
      return {
        campaignId,
        queryString,
        meta,
      };
    }),
    distinctUntilChanged((oldEvent, newEvent) => {
      if (oldEvent.campaignId === newEvent.campaignId && oldEvent.queryString === newEvent.queryString) {
        return !!newEvent.meta;
      }
      return false;
    }),
    concatMap(({ campaignId, queryString }) =>
      apiService.get(`/api/v1/campaigns/activate/campaigns/${campaignId}/contacts?${queryString}`, null, true).pipe(
        mergeMap((response: IContactsResponse) => [loadContactsSuccess(response)]),
        catchError(apiService.handleError(handlers.handleAnyError(loadContactsFail))),
      ),
    ),
  );

const syncContactsListOnPollEndedSideEffectEpic: Epic = (action$, state$) =>
  action$.pipe(
    ofType(pollContactsUploadingFileFinish),
    withLatestFrom(state$),
    filter(([, state]) => {
      const { mode } = getActivateModuleParams(state);
      return mode === ActivateModes.Editor;
    }),
    mergeMap(([, state]) => {
      const { campaignId } = getActivateModuleParams(state);
      return [resetStatusMetaData(), loadContactsRequest({ campaignId, reset: true })];
    }),
  );

export const contactsEpics = [
  loadContactsEpic,
  contactsQueryParamsSideEffectEpic,
  syncContactsListOnPollEndedSideEffectEpic,
];
