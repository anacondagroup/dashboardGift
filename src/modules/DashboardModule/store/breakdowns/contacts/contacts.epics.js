import { ofType } from 'redux-observable';
import { catchError, map, mergeMap, withLatestFrom, tap, debounce, takeUntil } from 'rxjs/operators/index';
import { of, timer } from 'rxjs';

import { CONTACTS_BREAKDOWN_LOAD_REQUEST, CONTACTS_DOWNLOAD_REPORT_REQUEST } from './contacts.types';
import {
  contactsLoadSuccess,
  contactsLoadFail,
  contactsDownloadReportSuccess,
  contactsDownloadReportFail,
} from './contacts.actions';
import { getReportFileName, buildApiUrl, buildReportUrl } from './contacts.selectors';

export const loadContactsBreakdownEpic = (action$, state$, { apiService }) =>
  action$.pipe(
    ofType(CONTACTS_BREAKDOWN_LOAD_REQUEST),
    debounce(() => timer(300)),
    map(action => buildApiUrl(action.payload)),
    mergeMap(url =>
      apiService.get(url).pipe(
        map(response => contactsLoadSuccess(response.recipients)),
        catchError(response => of(contactsLoadFail(response))),
        takeUntil(action$.ofType(CONTACTS_BREAKDOWN_LOAD_REQUEST)),
      ),
    ),
  );

export const downloadReportEpic = (action$, state$, { apiService, downloadFile }) =>
  action$.pipe(
    ofType(CONTACTS_DOWNLOAD_REPORT_REQUEST),
    map(action => [buildReportUrl(action.payload), action]),
    mergeMap(([url, action]) =>
      apiService.getFile(url).pipe(
        withLatestFrom(state$.pipe(map(getReportFileName(action.payload)))),
        tap(([blob, filename]) => {
          downloadFile(blob, filename);
        }),
        map(() => contactsDownloadReportSuccess()),
        catchError(() => contactsDownloadReportFail()),
      ),
    ),
  );

export const contactsEpic = [loadContactsBreakdownEpic, downloadReportEpic];
