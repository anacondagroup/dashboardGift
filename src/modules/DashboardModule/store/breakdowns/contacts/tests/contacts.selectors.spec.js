import { buildApiUrl, buildReportUrl, getReportFileName } from '../contacts.selectors';
import { fullStateMock } from '../../../../../../mocks/fullState.mock';

describe('contacts selector', () => {
  describe('getReportFileName', () => {
    it('Should build name without params', () => {
      const actual = getReportFileName({})(fullStateMock);
      expect(actual).toMatchSnapshot();
    });

    it('Should build name with all params', () => {
      const actual = getReportFileName({
        dateRangeFrom: '2019-03-22',
        dateRangeTo: '2019-03-29',
      })(fullStateMock);

      expect(actual).toMatchSnapshot();
    });
  });
  describe('buildApiUrl', () => {
    it('Should build url', () => {
      const actual = buildApiUrl({
        sort: 'name',
        sortDirection: 'asc',
        search: 'asd',
      });

      expect(actual).toMatchSnapshot();
    });
  });
  describe('buildReportUrl', () => {
    it('Should build url', () => {
      const actual = buildReportUrl({
        sort: 'name',
        sortDirection: 'asc',
        search: 'asd',
      });

      expect(actual).toMatchSnapshot();
    });
  });
});
