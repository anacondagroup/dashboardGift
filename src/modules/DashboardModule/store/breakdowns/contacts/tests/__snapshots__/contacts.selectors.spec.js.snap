// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contacts selector buildApiUrl Should build url 1`] = `"/enterprise/dashboard/breakdown/recipients?search=asd&sort_column=name&sort_direction=asc"`;

exports[`contacts selector buildReportUrl Should build url 1`] = `"/enterprise/report/recipient_breakdown?search=asd&sort_column=name&sort_direction=asc"`;

exports[`contacts selector getReportFileName Should build name with all params 1`] = `"Contact breakdown-2019-03-22-2019-03-29.xlsx"`;

exports[`contacts selector getReportFileName Should build name without params 1`] = `"Contact breakdown.xlsx"`;
